Extend inbound/outbound clipboard restriction

kenhys opened the pull request:
Closes: #2081 

This is a follow up #1298

Based on feedback, https://github.com/neutrinolabs/xrdp/issues/2081#issuecomment-992095542
RestrictOutboundClipboard configuration spec is extended and
RestrictInboundClipboard parameter is also supported.

```
  RestrictOutboundClipboard=file            # => Restricts file
  RestrictOutboundClipboard=file,text       # => Restricts file & text
  RestrictOutboundClipboard=true            # => Restricts text,file,image
```

NOTE: I'm not sure what is the best behavior about combined configuration case.
So, currently in the following case, `true` is ignored explicitly. 
```
  RestrictOutboundClipboard=true,text       # => Restricts text only

```

Supporting the above case (check priority) is out of scope in this PR.

metalefty commented:
Thanks! Regarding the case `true,text`, we need to discuss. Anyway, it is a good PR to start.

metalefty commented on sesman/sesman.ini.in:
```diff
@@ -21,7 +21,14 @@ TerminalServerAdmins=tsadmins
+; client is not pushed to the server.
+; In addition, you can control text/file/image transfer restrictions
+; respectively. It also accepts comma separated list such as text,file,image.
+RestrictInboundClipboard=false
```
TODO: Don't forget to edit man pages, too. It can be done later.

matt335672 replied:
Definitely leave this for now, along with unit tests.

kenhys replied:
Updated man page too.

metalefty commented on sesman/chansrv/chansrv_config.h:
```diff
@@ -39,6 +39,27 @@ struct config_chansrv
+    int restrict_inbound_clipboard;
+
+    /** RestrictInboundClipboard=text setting from sesman.ini */
+    int restrict_inbound_clipboard_text;
```
Is using bitmask here a good idea?

```
restrict_inbound_clipboard = TEXT | FILE;
```

metalefty replied:
Then we can create a new function like `g_text2bool()` to convert from string `text, file, image` to bitmask expression and vise versa.  We can also create unit tests for the conversion.

matt335672 replied:
Agreed, and covered above.

kenhys replied:
Thanks, I've changed to use bitmask.

matt335672 commented on sesman/chansrv/chansrv_config.c:
```diff
@@ -127,7 +128,47 @@ read_config_security(log_func_t logmsg,
+                if (g_find_property2bool(value, "image"))
+                {
+                    cfg->restrict_outbound_clipboard_image = 1;
+                }
```
This approach is simple and readable, but it doesn't address a current problem with the existing code which is that a mis-spelled keyword is not detected. Also, this part of the code can't easily be tested.

Like @metalefty, I think you want to use a bitmask to solve this, and an array of keywords and values which compute your value.

We have a structure in string calls.h called `struct bitmask_string` and a function called `g_bitmask_to_str()`. The function is a general way to convert a bitmask value into a string for logging.

What we could do with is a function called `g_str_to_bitmask()` which works the other way. Here's a signature:-

```c
/***
 * Converts a string containing a series of tokens to a bitmask.
 * @param str Input string
 * @param bitmask_string Array mapping tokens to bitmask values
 * @param delim Delimiter for tokens in str
 * @param[out] unrecognised Buffer for any unrecognised tokens
 * @param unrecognised_len Length of unrecognised including '`0';
 * @return bitmask value for recognised tokens
 */
int
g_str_to_bitmask(const char *str, const struct bitmask_string[],
                 char delim, char *unrecognised,
                 int unrecognised_len);
```

If the function encounters an unexpected token, it's added to unrecognised with a delimiter if necessary, which could be the passed-in delimiter or a space or whatever.

Then in chansrv_config.h you could have something like:-

```c
/* Define bitmask values for restricting the clipboard */
#define CLIP_RESTRICT_TEXT (1<<0)
#define CLIP_RESTRICT_FILE (1<<1)
#define CLIP_RESTRICT_IMAGE (1<<2)
#define CLIP_RESTRICT_ALL 0xffffffff
```

and in chansrv_config.c, rather than the above you could have:-

```c
/* Map clipboard strings into bitmask values */
const struct bitmask_string restrict_map[] = 
{
    { CLIP_RESTRICT_TEXT, "text"},
    { CLIP_RESTRICT_FILE, "file"},
    { CLIP_RESTRICT_IMAGE, "image"},
    { CLIP_RESTRICT_ALL, "all"},
    /* Compatibility values (do we want these?) */
    { CLIP_RESTRICT_ALL, "true"},
    { CLIP_RESTRICT_ALL, "on"},
    { CLIP_RESTRICT_ALL, "yes"},
    BITMASK_STRING_END_OF_LIST
};

. . .
     char unrecognised[256];
       if (g_strcasecmp(name, "RestrictOutboundClipboard") == 0)
       {
            cfg->restrict_outbound_clipboard = g_str_to_bitmask(value, restrict_map, ',', unrecognised, sizeof(unrecognised));
            if (unrecognised[0] != '\0')
            {
                LOG(LOG_LEVEL_WARNING,
                         "Unrecognised tokens parsing 'RestrictOutboundClipboard' %s",
                        unrecognised);
            }
        }
```

I think this is more likely to be useful for parsing other settings in files. It also means that adding an extra value just needs a line adding to the bitmask, and an extra line adding to `restrict_map` so it increases maintainability.

Also, (and this is important but we haven't covered it yet) we can add some test cases for this function to the test suite. Don't worry about this yet - let's get the basics in first.

The implementation of `g_str_to_bitmask()` can make a single pass through the string. You'll need to cope with whitespace in the string somehow. There's a function called `g_strtrim()` in the string library which can help you with this. Use `LOG(LOG_LEVEL_WARNING, 

Please come back to me if you think this isn't a reasonable way to proceed, or if I haven't explained myself very clearly. Also, there may well be better names to use than some of the ones I've picked above. I'm just trying to give you an idea rather than write your code for you.

kenhys replied:
Thanks, I've added `g_str_to_bitmask` and changed to use it.

matt335672 commented on common/string_calls.c:
```diff
@@ -143,6 +143,27 @@ g_text2bool(const char *s)
+    g_free(properties);
+    return 0;
+}
+
```
Doesn't cope with whitespace in the string (but see above)

matt335672 commented on sesman/chansrv/chansrv_config.c:
```diff
@@ -127,7 +128,47 @@ read_config_security(log_func_t logmsg,
+                if (g_find_property2bool(value, "image"))
+                {
+                    cfg->restrict_inbound_clipboard_image = 1;
+                }
```
See above. Same bitmask can be used.

matt335672 commented on sesman/config.c:
```diff
@@ -553,12 +600,19 @@ config_dump(struct config_sesman *config)
+    g_writeln("    AllowRootLogin:                   %d", sc->allow_root);
+    g_writeln("    MaxLoginRetry:                    %d", sc->login_retry);
+    g_writeln("    AlwaysGroupCheck:                 %d", sc->ts_always_group_check);
+    g_writeln("    RestrictOutboundClipboard:        %d", sc->restrict_outbound_clipboard);
```
Can use the existing `g_bitmask_to_str()` o display bitmap values here.

kenhys replied:
Fixed to use it.

matt335672 commented on sesman/chansrv/clipboard.c:
```diff
@@ -1999,20 +1999,40 @@ clipboard_event_selection_notify(XEvent *xevent)
+                        {
+                            clipboard_send_data_response_for_file(g_clip_s2c.data,
+                                                                  g_clip_s2c.total_bytes);
+                        }
```
I'll admit to not looking to closely at this code for now, as I've spent a while on the config stuff which I think is important to get right. We can cover this all later.

In general however, you probably want to use `LOG()` instead of `LOG_DEVEL()`. See https://github.com/neutrinolabs/xrdp/wiki/Logging for more info on the logging and please ask if you have any questions.

kenhys replied:
https://github.com/neutrinolabs/xrdp/blob/5a11b698ef0a501d01daa2addf63b501d79935c5/sesman/chansrv/clipboard.c#L2535-L2538

It seems that the existing implementation uses `LOG_DEVEL`.
`LOG` is better for it?

kenhys replied:
I've changed to use `LOG` in those context.

matt335672 replied:
I think LOG() is better here, but sometimes this can be a matter of opinion rather than being obvious.

As a general rule, if a system manager would find it useful use a LOG(), whereas if a developer/maintainer would find it useful, use LOG_DEVEL().

kenhys replied:
I've created a separate PR above case only.
https://github.com/neutrinolabs/xrdp/pull/2088

matt335672 reviewed (commented):
@kenhys - first of all, thanks for taking the time to submit this PR. I think it's a great idea, and I'll happily give to help to get this in to the product.

kenhys committed:
Add g_str_to_bitmask utility function

It should be used for comma separated configuration to bitmask.

e.g. RestrictOutboundClipboard = text, file, image

kenhys committed:
Extend In/Outbound text,file,image restriction respectively

It supports the extended configurations for sesman.ini:

Before:

  [Security]
  RestrictOutboundClipboard=true or false

After:

  [Security]
  RestrictInboundClipboard=[true or false | text or file or image | comma separated list]
  RestrictOutboundClipboard=[true or false | text or file or image | comma separated list]

Above configuration is disabled by default (false)
And it can be specified comma separated list like this:.

  RestrictInboundClipboard=file, image
  RestrictOutboundClipboard=text, file, image

Note that if RestrictOutboundClipboard=true,file is set,
file is ignored and it is treated as RestrictOutboundClipboard=true

It is same for RestrictInboundClipboard.

kenhys committed:
Add sesman.ini new text/file/image restriction settings

RestrictInboundClipboard is added.

Then, RestrictOutboundClipboard/RestrictInboundClipboard configuration
is extended to accept comma separated list.

  * RestrictOutboundClipboard=none
  * RestrictOutboundClipboard=text
  * RestrictOutboundClipboard=file
  * RestrictOutboundClipboard=image
  * RestrictOutboundClipboard=all
  * RestrictOutboundClipboard=text, image, file

For compatibility, the following configuration is also
accepted (alias)

  * RestrictOutboundClipboard=true
  * RestrictOutboundClipboard=false
  * RestrictOutboundClipboard=yes

kenhys committed:
Block outbound clipboard text/image/file respectively

RestrictOutboundClipboard kills all of test/file/image
transfer via clipboard.

For controlling each content type behavior,
clipboard_xevent is not appropriate place to block respectively.

Instead, in clipboard_event_selection_notify, these media type
will be blocked which depends on the following configurations in
sesman.ini [Security] section.

  * RestrictOutboundClipboard=text
  * RestrictOutboundClipboard=file
  * RestrictOutboundClipboard=image

You can also set comma separated list

  * RestrictOutboundClipboard=text, file, image

kenhys committed:
Block inbound clipboard text/image/file respectively

Disable clipboard_event_selection_request call is overkill for
blocking text/image/file purpose.
For example, it breaks existing behavior (slow response from gedit,
gimp as a side effects)

Instead, in clipboard_event_selection_request, these media format will
be blocked respectively which depends on the following configurations
in sesman.ini [Security] section.

  * RestrictInboundClipboard=text
  * RestrictInboundClipboard=file
  * RestrictInboundClipboard=image

You can also set comma separated list.

  * RestrictInboundClipboard=text,file,image

kenhys committed:
Update sesman.ini.5 explanation about RestrictOutboundClipboard,RestrictOutboundClipboard

RestrictOutboundClipboard,RestrictOutboundClipboard are extended to
accept text,file,image configuration value.

kenhys commented:
* Implement g_str_to_bitmask
* Changed to use g_str_to_bitmask
* Updated sesman.ini
* Updated sesman.ini.5 manpage
* Updated inbound/outbound patch

matt335672 commented on common/string_calls.c:
```diff
@@ -986,3 +986,56 @@ g_bitmask_to_str(int bitmask, const struct bitmask_string bitdefs[],
+    g_memset(unrecognised, 0, unrecognised_len);
+    if (!str || !p)
+    {
+        return 0;
```
Two problems here:-
1) If `str` is NULL, then `properties` is NULL, and the `strtok()` will crash before we get to this line.
2) Memory leak here on the return at 1001 if str is "" rather than NULL. If str is "", the strtok returns NULL in p. However, the string copy from 994 has still been made.

Suggest moving the guard on `str` before the `g_strdup()`. Also, the guard on `p` can be removed by replacing the `do { } while(p);` loop below with a `while(p != NULL) { }` loop.

kenhys replied:
Fixed.

matt335672 commented on common/string_calls.c:
```diff
@@ -986,3 +986,56 @@ g_bitmask_to_str(int bitmask, const struct bitmask_string bitdefs[],
+        g_strtrim(p, 3);
+        const struct bitmask_string *b;
+        int found = 0;
+        for (b = &bitdefs[0] ; b->mask != 0; ++b)
```
Suggest using test of `p->str != NULL` rather than `b->mask != 0`. The user may use a mask of zero for options that don't do anything, and if so all the values underneath the zero will be ignored.

kenhys replied:
Fixed. and added test case for it.

matt335672 commented on common/string_calls.c:
```diff
@@ -986,3 +986,56 @@ g_bitmask_to_str(int bitmask, const struct bitmask_string bitdefs[],
+            if (0 == g_strcasecmp(p, b->str))
+            {
+                mask |= b->mask;
+                found = 1;
```
Can `break` here to exit loop early for found values.

kenhys replied:
Fixed.

matt335672 commented on common/string_calls.c:
```diff
@@ -986,3 +986,56 @@ g_bitmask_to_str(int bitmask, const struct bitmask_string bitdefs[],
+                {
+                    g_strncat(unrecognised, ",", 1);
+                    length += 1;
+                }
```
When adding the delimiter here, suggest using the same one the user passed in (i.e.):-

```c
                if (length + g_strlen(delim) < unrecognised_len)
                {
                    g_strcpy(unrecognised + length, delim);
                    length += g_strlen(delim);
                }
```

The g_strcpy() is safe to use here, as the previous line checks that it is.

kenhys replied:
Fixed.

matt335672 commented on common/string_calls.c:
```diff
@@ -986,3 +986,56 @@ g_bitmask_to_str(int bitmask, const struct bitmask_string bitdefs[],
+            {
+                g_strncat(unrecognised, p, g_strlen(p));
+                length += g_strlen(p);
+                unrecognised[length] = '\0';
```
As above, suggest:-

```c
    g_strcpy(unrecognised + length, p);
    length += g_strlen(p);
```

The explicit terminator is unnecessary as `g_strcpy()` adds it automatically.

kenhys replied:
Fixed.

matt335672 commented on common/string_calls.c:
```diff
@@ -986,3 +986,56 @@ g_bitmask_to_str(int bitmask, const struct bitmask_string bitdefs[],
+                unrecognised[length] = '\0';
+            }
+        }
+        p = strtok(NULL, ",");
```
Replace `","` with `delim`. This isn't picked up in the unit tests, but can be with  small change I've mentioned below.

kenhys replied:
Fixed.

kenhys replied:
Added test case.

matt335672 commented on common/string_calls.c:
```diff
@@ -986,3 +986,56 @@ g_bitmask_to_str(int bitmask, const struct bitmask_string bitdefs[],
+    char *p = strtok(properties, delim);
+    int mask = 0;
+
+    g_memset(unrecognised, 0, unrecognised_len);
```
Don't need to set the whole buffer to zero here. The following code is good enough:-

```c
if (unrecognised_len < 1)
{
    /* No good solution for this */
    return 0;
}
unrecognised[unrecognised_len] = '\0';
```

kenhys replied:
Thanks, `unrecognised[0] = '\0'` is also needed.

matt335672 replied:
Good catch - we don't need `unrecognised[unrecognised_len] = '\0';` - that's my mistake.

matt335672 commented on docs/man/sesman.ini.5.in:
```diff
@@ -221,10 +221,20 @@ login for all users is enabled.
 
 .TP
-\fBRestrictOutboundClipboard\fR=\fI[true|false]\fR
+\fBRestrictOutboundClipboard\fR=\fI[true|false|text|file|image]\fR
```
No mention of 'all' either here or below.

I think we need to decide whether the 'all bits set' case is best described by 'all' or 'true'. I don't really have an opinion on this, one way or another - I'm happy with either. @metalefty - what do you think?

metalefty replied:
I'm thinking in this way. 

Options we'll announce officially:
* all
* text
* file
* image
* none

Hidden options for compatibility (warn in the log if needed):
* true (an alias of all)
* yes (an alias of all)
* false (an alias of none)

kenhys replied:
Fixed.

matt335672 commented on sesman/chansrv/chansrv_config.c:
```diff
@@ -125,9 +141,26 @@ read_config_security(log_func_t logmsg,
         if (g_strcasecmp(name, "RestrictOutboundClipboard") == 0)
         {
-            cfg->restrict_outbound_clipboard = g_text2bool(value);
+            cfg->restrict_outbound_clipboard = g_str_to_bitmask(value, clip_restrict_map, ",", unrecognised, sizeof(unrecognised));
```
Coding standard (see [wiki](/neutrinolabs/xrdp/wiki/Coding-Style)) asks for lines to be 80 characters if possible. Suggest something like:-

```c
            cfg->restrict_outbound_clipboard =
                g_str_to_bitmask(value, clip_restrict_map, ",",
                                 unrecognised, sizeof(unrecognised));
```

Similar comments apply to lines 151, 157, 335, 348

kenhys replied:
Fixed.

matt335672 commented on sesman/chansrv/chansrv_config.c:
```diff
@@ -211,6 +244,7 @@ new_config(void)
         cfg->fuse_mount_name = fuse_mount_name;
         cfg->file_umask = DEFAULT_FILE_UMASK;
         cfg->use_nautilus3_flist_format = DEFAULT_USE_NAUTILUS3_FLIST_FORMAT;
+        cfg->restrict_inbound_clipboard = DEFAULT_RESTRICT_INBOUND_CLIPBOARD;
```
Purely stylistically, I'd suggest keeping `cfg->restrict_outbound_clipboard` and `cfg->restrict_inbound_clipboard` together, as they're doing similar jobs. This applies to the header file too.

kenhys replied:
Fixed.

matt335672 commented on sesman/config.c:
```diff
@@ -231,7 +252,25 @@ config_read_security(int file, struct config_security *sc,
         {
-            sc->restrict_outbound_clipboard = g_text2bool((char *)list_get_item(param_v, i));
+            char unrecognised[256];
+            sc->restrict_outbound_clipboard = g_str_to_bitmask((const char *)list_get_item(param_v, i), clip_restrict_map, ",", unrecognised, sizeof(unrecognised));
```
Line length over 80 (see previous comment).

kenhys replied:
Fixed.

matt335672 commented on sesman/chansrv/chansrv_config.c:
```diff
@@ -286,10 +320,34 @@ config_dump(struct config_chansrv *config)
+    }
+    else
+    {
+        g_bitmask_to_str(config->restrict_outbound_clipboard, clip_restrict_map, ',', buf, 256);
```
suggest `sizeof(buf)` rather than `256` for maintenance reasons. Same comment for line 348.

kenhys replied:
Fixed.

matt335672 commented on tests/common/test_string_calls.c:
```diff
@@ -323,6 +323,127 @@ START_TEST(test_bm2str__overflow_some_bits_undefined)
+    ck_assert_int_eq(rv, bitmask);
+}
+END_TEST
+
```
Suggest here adding a BIT_2 value as well. If you do this, it catches the error in the function where you have a hard-coded "," in the second `strtok()` call.

kenhys replied:
Fixed.

matt335672 reviewed (commented):
@kenhys - this is great work!

I've still not reviewed clipboard.c, as I'm out of time today and I though it was worth concentrating on the string functions. I'll get to it when I can.

Really pleased about the module tests too.

metalefty commented:
BTW, we'll release v0.9.18 soon. I'm fine with getting this in the next release but we need triage because we have some other ongoing PRs. 

I think #2068 has a higher priority so this PR might be merged right after the 0.9.18 release.

matt335672 commented:
@kenhys - I've had to concentrate on an important issue for v0.9.18 today, and I've got family commitments for the rest of this week. I'll have to pick this up next week now. Thanks for all the work you're putting in to this.

kenhys commented:
Thank you for sharing the status!

No problem, no need to rush into it. I'll wait for your reviews.

matt335672 commented on common/string_calls.c:
```diff
@@ -986,3 +986,72 @@ g_bitmask_to_str(int bitmask, const struct bitmask_string bitdefs[],
+    {
+        return 0;
+    }
+    unrecognised[unrecognised_len] = '\0';
```
Please remove this line as it writes a terminator past the end of the passed-in buffer. I know I asked for this line in a previous comment (sorry, again), but as you yourself pointed out it is incorrect.

kenhys replied:
Thanks, fixed.

matt335672 commented on common/string_calls.c:
```diff
@@ -986,3 +986,72 @@ g_bitmask_to_str(int bitmask, const struct bitmask_string bitdefs[],
+        return 0;
+    }
+    unrecognised[unrecognised_len] = '\0';
+    properties = g_strdup(str);
```
`g_strdup()` might return NULL - this should be checked.

kenhys replied:
Fixed.

matt335672 commented on docs/man/sesman.ini.5.in:
```diff
@@ -221,10 +221,62 @@ login for all users is enabled.
-\fBRestrictOutboundClipboard\fR=\fI[true|false]\fR
-If set to \fB1\fR, \fBtrue\fR or \fByes\fR, will restrict the clipboard
+\fBRestrictOutboundClipboard\fR=\fI[all|none|text|file|image]\fR
+If set to \fBall\fR, \fBtrue\fR or \fByes\fR, will restrict the clipboard
```
Suggest not mentioning 'true' or 'yes' in this line, as you've done that nicely below.

kenhys replied:
Fixed.

matt335672 commented on docs/man/sesman.ini.5.in:
```diff
@@ -221,10 +221,62 @@ login for all users is enabled.
+
+.TP
+\fBRestrictInboundClipboard\fR=\fI[none|all|text|file|image]\fR
+If set to \fBall\fR, \fBtrue\fR or \fByes\fR, will restrict the clipboard
```
Suggest not mentioning 'true' or 'yes' in this line, as you've done that nicely below.

kenhys replied:
Fixed.

matt335672 commented on tests/common/test_string_calls.c:
```diff
@@ -323,6 +323,312 @@ START_TEST(test_bm2str__overflow_some_bits_undefined)
+START_TEST(test_str2bm__zero_buffer)
+{
+    int rv;
+    char buff[0];
```
Arrays of size zero are not allowed by the [C99 standard](http://www.open-std.org/jtc1/sc22/wg14/www/docs/n1256.pdf) section 6.7.5.2, although gcc allows them as an extension. The problem with them is that when you access the array (at line 423), the access is undefined.

Suggest you make this `char buff[1]`, replace the `sizeof(buff)` at line 421 with '0', and remove the access to the buffer at line 423. You'll still be testing that zero is returned for an invalid buffer.

kenhys replied:
Fixed.

matt335672 commented on common/string_calls.c:
```diff
@@ -986,3 +986,72 @@ g_bitmask_to_str(int bitmask, const struct bitmask_string bitdefs[],
+                /* adding ",property" */
+                if (length + g_strlen(p) + 1 < unrecognised_len)
+                {
+                    g_strncat(unrecognised, ",", 1);
```
Better as `unrecognised[length] = delim[0];`, so you're avoiding an unnecessary string traversal, and also using the first delimiter the user passes in (which may not be ',').

kenhys replied:
Fixed.

matt335672 commented on common/string_calls.c:
```diff
@@ -986,3 +986,72 @@ g_bitmask_to_str(int bitmask, const struct bitmask_string bitdefs[],
+                }
+            }
+            else
+            {
```
Unnecessary indentation. Replace:-

```c
else
{
    if (...)
    {
    }
}
```

with :-

```c
else if (...)
{
}
```

kenhys replied:
Fixed.

matt335672 commented on common/string_calls.c:
```diff
@@ -986,3 +986,72 @@ g_bitmask_to_str(int bitmask, const struct bitmask_string bitdefs[],
+            }
+            else
+            {
+                if (g_strlen(p) + 1 < unrecognised_len)
```
`+1` is not necessary and can be removed.

kenhys replied:
Fixed.

matt335672 commented on sesman/chansrv/clipboard.c:
```diff
@@ -2243,19 +2350,47 @@ clipboard_event_selection_request(XEvent *xevent)
+                    g_cfg->restrict_inbound_clipboard == CLIP_RESTRICT_ALL ? "all" : "file",
+                    lxev->target == XA_STRING ? "XA_STRING" : "UTF8_STRING");
+                clipboard_refuse_selection(lxev);
+            }
```
Suggest simplifying the above to:-

```c
            if (g_cfg->restrict_inbound_clipboard & CLIP_RESTRICT_FILE)
            {
                LOG(LOG_LEVEL_DEBUG,
                    "inbound clipboard %s is restricted because of config",
                    lxev->target == XA_STRING ? "XA_STRING" : "UTF8_STRING");
                clipboard_refuse_selection(lxev);
            }
```

we've already logged the parameter setting as 'file' or 'all', so there's no need to be more explicit here.

kenhys replied:
Fixed.

matt335672 commented on sesman/chansrv/clipboard.c:
```diff
@@ -2243,19 +2350,47 @@ clipboard_event_selection_request(XEvent *xevent)
+                    g_cfg->restrict_inbound_clipboard == CLIP_RESTRICT_ALL ? "all" : "text",
+                    lxev->target == XA_STRING ? "XA_STRING" : "UTF8_STRING");
+                clipboard_refuse_selection(lxev);
+            }
```
Suggest removing test against `CLIPBOARD_RESTRICT_ALL` and simplifying log message (previous comment)

kenhys replied:
Fixed.

matt335672 commented on sesman/chansrv/clipboard.c:
```diff
@@ -2265,47 +2400,123 @@ clipboard_event_selection_request(XEvent *xevent)
+                    g_cfg->restrict_inbound_clipboard == CLIP_RESTRICT_ALL ? "all" : "image");
+                clipboard_refuse_selection(lxev);
+            }
+            else
```
Suggest removing test against `CLIPBOARD_RESTRICT_ALL` and simplifying log message (previous comment)

kenhys replied:
Fixed.

matt335672 commented on sesman/chansrv/clipboard.c:
```diff
@@ -2265,47 +2400,123 @@ clipboard_event_selection_request(XEvent *xevent)
+                "inbound clipboard(%s) image/bmp is restricted because of config",
+                g_cfg->restrict_inbound_clipboard == CLIP_RESTRICT_ALL ? "all" : "image");
+            clipboard_refuse_selection(lxev);
+        }
```
Suggest removing test against `CLIPBOARD_RESTRICT_ALL` and simplifying log message (previous comment)

kenhys replied:
Fixed.

matt335672 commented on sesman/chansrv/clipboard.c:
```diff
@@ -2265,47 +2400,123 @@ clipboard_event_selection_request(XEvent *xevent)
+                LOG(LOG_LEVEL_DEBUG,
+                    "inbound clipboard(%s) text/uri-list is restricted because of config",
+                    g_cfg->restrict_inbound_clipboard == CLIP_RESTRICT_ALL ? "all" : "file");
+                clipboard_refuse_selection(lxev);
```
Suggest removing test against `CLIPBOARD_RESTRICT_ALL` and simplifying log message (previous comment)

kenhys replied:
Fixed.

matt335672 commented on sesman/chansrv/clipboard.c:
```diff
@@ -2265,47 +2400,123 @@ clipboard_event_selection_request(XEvent *xevent)
+            LOG(LOG_LEVEL_DEBUG,
+                "inbound clipboard(%s) text/uri-list is restricted because of config",
+                g_cfg->restrict_inbound_clipboard == CLIP_RESTRICT_ALL ? "all" : "file");
+            clipboard_refuse_selection(lxev);
```
Suggest removing test against `CLIPBOARD_RESTRICT_ALL` and simplifying log message (previous comment)

kenhys replied:
Fixed.

matt335672 commented on sesman/chansrv/clipboard.c:
```diff
@@ -2265,47 +2400,123 @@ clipboard_event_selection_request(XEvent *xevent)
+                LOG(LOG_LEVEL_DEBUG,
+                    "inbound clipboard(%s) x-special/gnome-copied-files converted is restricted because of config",
+                    g_cfg->restrict_inbound_clipboard == CLIP_RESTRICT_ALL ? "all" : "file");
+                clipboard_refuse_selection(lxev);
```
Suggest removing test against `CLIPBOARD_RESTRICT_ALL` and simplifying log message (previous comment)

kenhys replied:
Fixed.

matt335672 commented on sesman/chansrv/clipboard.c:
```diff
@@ -2265,47 +2400,123 @@ clipboard_event_selection_request(XEvent *xevent)
+            LOG(LOG_LEVEL_DEBUG,
+                "inbound clipboard(%s) x-special/gnome-copied-files is restricted because of config",
+                g_cfg->restrict_inbound_clipboard == CLIP_RESTRICT_ALL ? "all" : "file");
+            clipboard_refuse_selection(lxev);
```
Suggest removing test against `CLIPBOARD_RESTRICT_ALL` and simplifying log message (previous comment)

kenhys replied:
Fixed.

matt335672 commented on sesman/chansrv/clipboard.c:
```diff
@@ -2527,7 +2738,7 @@ clipboard_xevent(void *xevent)
     {
         case SelectionNotify:
-            if (g_cfg->restrict_outbound_clipboard == 0)
+            if (g_cfg->restrict_outbound_clipboard != CLIP_RESTRICT_ALL)
```
Suggest replacing:-

```c
            if (g_cfg->restrict_outbound_clipboard != CLIP_RESTRICT_ALL)
            {
                clipboard_event_selection_notify(lxevent);
            }
            else
            {
                LOG_DEVEL(LOG_LEVEL_DEBUG, "outbound clipboard is restricted because of config");
                return 1;
            }
```

with:-

```c
                clipboard_event_selection_notify(lxevent);
```

The changes you've made to `clipboard_event_selection_notify()` will handle this perfectly well.

kenhys replied:
Fixed.

matt335672 reviewed (commented):
@kenhys - I've finally found time to fully review this.

In addition to these, I've got one related to advertising formats. I'll add that below.

matt335672 commented:
In addition to the review above, some extra code needs to be added to `clipboard_event_selection_request()` which relates to the formats which we advertise over X.

At the moment, if `RestrictOutboundClipboard=none` (for example), we're still advertising formats we won't provide to the X server. You can see this by:-
- setting `RestrictOutboundClipboard=none`
- copying some text on the Windows client
- Trying to paste on the Linux side by using a menu.

On GNOME I get this context menu, containing a 'Paste' option:-
![PasteAllowed](https://user-images.githubusercontent.com/30179339/147351606-c158b9ce-09e6-4632-b03f-7708a760ec84.png)

When I select 'Paste', nothing happens, as you'd expect from the code.

What I think should happen is that if the text clipboard is restricted I should get this menu instead, without 'Paste' being offered as an option:-
![PasteNotAllowed](https://user-images.githubusercontent.com/30179339/147351677-d33a7f8c-6637-4bdf-a33a-76a7140b8642.png)

Here's a patch which I think will fix this:-

```diff
--- a/sesman/chansrv/clipboard.c
+++ b/sesman/chansrv/clipboard.c
@@ -2294,16 +2294,24 @@ clipboard_event_selection_request(XEvent *xevent)
         atom_buf[0] = g_targets_atom;
         atom_buf[1] = g_timestamp_atom;
         atom_buf[2] = g_multiple_atom;
-        atom_buf[3] = XA_STRING;
-        atom_buf[4] = g_utf8_atom;
-        atom_count = 5;
-        if (clipboard_find_format_id(CB_FORMAT_DIB) >= 0)
+        atom_count = 3;
+        if ((g_cfg->restrict_inbound_clipboard & CLIP_RESTRICT_TEXT) == 0)
+        {
+            atom_buf[atom_count] = XA_STRING;
+            atom_count++;
+            atom_buf[atom_count] = g_utf8_atom;
+            atom_count++;
+        }
+        if (clipboard_find_format_id(CB_FORMAT_DIB) >= 0 &&
+            (g_cfg->restrict_inbound_clipboard & CLIP_RESTRICT_IMAGE) == 0)
         {
             LOG_DEVEL(LOG_LEVEL_DEBUG, "  reporting image/bmp");
             atom_buf[atom_count] = g_image_bmp_atom;
             atom_count++;
         }
-        if (clipboard_find_format_id(g_file_format_id) >= 0)
+        if (clipboard_find_format_id(g_file_format_id) >= 0 &&
+            (g_cfg->restrict_inbound_clipboard & CLIP_RESTRICT_FILE) == 0)
+
         {
             LOG_DEVEL(LOG_LEVEL_DEBUG, "  reporting text/uri-list");
             atom_buf[atom_count] = g_file_atom1;
```

Happy to discuss either this, or anything else in my review.

kenhys commented:
@matt335672 Thanks.

I've tried suggested patch, 

* For c2s text, paste context menu is disabled. (expected)
* For c2s file, thunar context menu is still enabled, but nothing happen when paste is clicked (without suggested patch, it shows error dialog)
* For c2s image, gimp 'paste' menu is still enabled, and it shows error dialog when it is clicked.

As for c2s text, suggested patch works well, but for c2s file and image, it seems that it lacks a something.

matt335672 commented:
@kenhys - apologies again for not getting back to you sooner. Now we've got 0.9.18 out I'll have some time to look at this (and my broken patch) later in the week.

matt335672 commented:
@kenhys - I've had another look at the patch I suggested, and it seems OK to me.

One possible point of confusion is that when I suggested it [here](/neutrinolabs/xrdp/pull/2082#issuecomment-1000817785) I was talking about `RestrictOutboundClipboard`, which is incorrect. It should be `RestrictInboundClipboard` (i.e. the c2s direction).

Here's the patch again with new line numbers after your changes:-

```diff
--- a/sesman/chansrv/clipboard.c
+++ b/sesman/chansrv/clipboard.c
@@ -2294,16 +2294,23 @@ clipboard_event_selection_request(XEvent *xevent)
         atom_buf[0] = g_targets_atom;
         atom_buf[1] = g_timestamp_atom;
         atom_buf[2] = g_multiple_atom;
-        atom_buf[3] = XA_STRING;
-        atom_buf[4] = g_utf8_atom;
-        atom_count = 5;
-        if (clipboard_find_format_id(CB_FORMAT_DIB) >= 0)
+        atom_count = 3;
+        if ((g_cfg->restrict_inbound_clipboard & CLIP_RESTRICT_TEXT) == 0)
+        {
+            atom_buf[atom_count] = XA_STRING;
+            atom_count++;
+            atom_buf[atom_count] = g_utf8_atom;
+            atom_count++;
+        }
+        if (clipboard_find_format_id(CB_FORMAT_DIB) >= 0 &&
+                (g_cfg->restrict_inbound_clipboard & CLIP_RESTRICT_IMAGE) == 0)
         {
             LOG_DEVEL(LOG_LEVEL_DEBUG, "  reporting image/bmp");
             atom_buf[atom_count] = g_image_bmp_atom;
             atom_count++;
         }
-        if (clipboard_find_format_id(g_file_format_id) >= 0)
+        if (clipboard_find_format_id(g_file_format_id) >= 0 &&
+                (g_cfg->restrict_inbound_clipboard & CLIP_RESTRICT_FILE) == 0)
         {
             LOG_DEVEL(LOG_LEVEL_DEBUG, "  reporting text/uri-list");
             atom_buf[atom_count] = g_file_atom1;
```

You should be able to apply this directly from the xrdp top-level directory with `patch -p1 <[patchfile]`.

With that patch on top of your code I've got the following test results, which I'm happy with. Please have a look at these and let me know what you think, or if I haven't explained anything well enough.

I've tested in both directions, although the patch above only affects c2s.

# RestrictInboundClipboard = none
## Text
1) Copy text in Windows Notepad
2) Check TARGETS atom:-
```
$ xclip -o -t TARGETS -selection clipboard
TARGETS
TIMESTAMP
MULTIPLE
STRING
UTF8_STRING
```
3) Right-click menu in terminal; 'Paste' is allowed
4) Paste text - all OK
## File
1) Copy file in Windows Explorer
2) Check TARGETS atom:-
```
$ xclip -o -t TARGETS -selection clipboard
TARGETS
TIMESTAMP
MULTIPLE
STRING
UTF8_STRING
text/uri-list
x-special/gnome-copied-files
```
3) Right-click in Thunar; 'Paste' is allowed
4) Paste file - all OK
## Image
1) Copy image with Windows snipping tool
2) Check TARGETS atom:-
```
$ xclip -o -t TARGETS -selection clipboard
TARGETS
TIMESTAMP
MULTIPLE
STRING
UTF8_STRING
image/bmp
```
3) Paste in GIMP; image comes over OK

# RestrictInboundClipboard = all
## Text
1) Copy text in Windows Notepad
2) Check TARGETS atom:-
```
$ xclip -o -t TARGETS -selection clipboard
TARGETS
TIMESTAMP
MULTIPLE
```
3) Right-click menu in terminal; 'Paste' is not allowed and cannot be selected
## File
1) Copy file in Windows Explorer
2) Check TARGETS atom:-
```
$ xclip -o -t TARGETS -selection clipboard
TARGETS
TIMESTAMP
MULTIPLE
```
3) Right-click in Thunar; 'Paste' is not allowed and cannot be selected
## Image
1) Copy image with Windows snipping tool
2) Check TARGETS atom:-
```
$ xclip -o -t TARGETS -selection clipboard
TARGETS
TIMESTAMP
MULTIPLE
```
3) Paste in GIMP; dialog appears
![gimp](https://user-images.githubusercontent.com/30179339/149344352-7655b49e-b61a-4ab7-bfc9-a386276e8a61.png)

The only odd thing is the Gimp dialog. However, I have never seen 'Paste' greyed out in Gimp, and `xclip` seems to suggest everything is OK so I think that's down to Gimp.

# RestrictOutboundClipboard = none
For these I still check the TARGETS atom, but this is now set by XFCE when the copy is made, and not by xrdp.
## Text
1) Copy text in terminal
2) Check TARGETS atom:-
```
$ xclip -o -t TARGETS -selection clipboard
TIMESTAMP
TARGETS
MULTIPLE
SAVE_TARGETS
UTF8_STRING
COMPOUND_TEXT
TEXT
STRING
text/plain;charset=utf-8
text/plain
```
3) Right-click menu in Windows notepad; 'Paste' is allowed
4) Paste text - all OK
## File
1) Copy file in Thunar.
2) Check TARGETS atom:-
```
$ xclip -o -t TARGETS -selection clipboard
TIMESTAMP
TARGETS
MULTIPLE
text/uri-list
x-special/gnome-copied-files
UTF8_STRING
```
3) Right-click in Windows Explorer; 'Paste' is allowed
4) Paste file - all OK
## Image
1) Copy image in Gimp with select-all/copy
2) Check TARGETS atom:-
```
$ xclip -o -t TARGETS -selection clipboard
TIMESTAMP
TARGETS
MULTIPLE
SAVE_TARGETS
image/png
image/bmp
image/x-bmp
image/x-MS-bmp
image/x-icon
image/x-ico
image/x-win-bitmap
image/vnd.microsoft.icon
application/ico
image/ico
image/icon
text/ico
image/tiff
image/jpeg
```
3) Paste in Windows Paint; image comes over OK

# RestrictOutboundClipboard = all
For these I still check the TARGETS atom, but this is now set by XFCE when the copy is made, and not by xrdp.
## Text
1) Copy text in terminal
2) Check TARGETS atom:-
```
$ xclip -o -t TARGETS -selection clipboard
TIMESTAMP
TARGETS
MULTIPLE
SAVE_TARGETS
UTF8_STRING
COMPOUND_TEXT
TEXT
STRING
text/plain;charset=utf-8
text/plain
```
3) Right-click menu in Windows notepad; 'Paste' is not allowed and cannot be selected
## File
1) Copy file in Thunar.
2) Check TARGETS atom:-
```
$ xclip -o -t TARGETS -selection clipboard
TIMESTAMP
TARGETS
MULTIPLE
text/uri-list
x-special/gnome-copied-files
UTF8_STRING
```
3) Right-click in Windows Explorer; 'Paste' is not allowed and cannot be selected
## Image
1) Copy image in Gimp with select-all/copy
2) Check TARGETS atom:-
```
$ xclip -o -t TARGETS -selection clipboard
TIMESTAMP
TARGETS
MULTIPLE
SAVE_TARGETS
image/png
image/bmp
image/x-bmp
image/x-MS-bmp
image/x-icon
image/x-ico
image/x-win-bitmap
image/vnd.microsoft.icon
application/ico
image/ico
image/icon
text/ico
image/tiff
image/jpeg
```
3) Paste in Windows Paint is not allowed and cannot be selected.

matt335672 reviewed (approved):
@kenhys - I think this is good work.

I must apologise again for taking a while to get to this. I'm aware you've responded very quickly to my reviews and I've taken some time to get back to you after yours.

Thanks again for contributing to xrdp.

kenhys commented:
@matt335672 

Thank you for the verification.
I'm happy to hear that.
Then could you commit the above additional patch and merge this PR, please?

kenhys commented:
I've just rebased with devel branch to keep merge history simple.

metalefty commented:
@kenhys @matt335672 Thanks you both. Much improved clipboard restriction.

matt335672 committed:
Only advertise X11 clip formats we can supply

matt335672 commented:
As requested, patch added and merge made.

Thanks again, @kenhys
